### ISSUE TITLE
Screen flickering when using neovim in tmux

### ISSUE THREAD

**Issue by mlitwiniuk** [neovim/neovim#2528]
Whenever I run neovim in tmux it's no longer behaves smoothly, screen starts to be refreshed in a strange way.
Here's video of that behaviour: http://up.prograils.com/ml/nvim_tmux.mp4

Config files:
[nvimrc](https://github.com/mlitwiniuk/dotfiles/blob/master/vimrc)
[tmux.conf](https://github.com/mlitwiniuk/dotfiles/blob/master/tmux.conf)

I really have no idea what the hell is going on there.

**Comment by Pyrohh** [comment]
I experience a similar issue, but it's much less prevalent after disabling `lazyredraw`, although it still happens.

**Comment by mhinz** [comment]
Hmm, I can't reproduce it. Do you experience the same with `nvim -u NONE`?

It seems that you're suggesting that it used to work. Do you install nvim by package manager or compile from scratch? It would be great if you could `git bisect` this issue in the latter case.

**Comment by mlitwiniuk** [comment]
`nvim -u NONE` works fine. I'll start with finding config, that causes flickering

**Comment by crepererum** [comment]
Just as a general hint: always ensure that `$TERM` is set to `screen[.*]` when using neovim and tmux. Sometimes you overwrite this in your shell config and therefor break a lot of things (also some keystrokes like HOME and END). So `TERM=xterm` is not working correctly in combination with tmux (at least not for me), which seems to be legit.

**Comment by mlitwiniuk** [comment]
When enabling gutter (ie via `set rnu`) problem starts to be visible even with `nvim -u NONE`. Disabling `colorcolumn` helps a little. Still thou - this is far from beeing perfect.

**Comment by justinmk** [comment]
Redraw/flicker is an ongoing effort, this issue isn't narrow enough.

**Comment by Pyrohh** [comment]
@justinmk I don't see another issue about tmux related flickering, which is the only place I'm experiencing this.

**Comment by Pyrohh** [comment]
setup information:
- `NVIM 0.0.0-alpha+201507262350 (compiled Jul 27 2015 13:18:07)`
- st 0.6, used with the arguments `-f terminus-13`
- xterm (`xterm -v` is  `XTerm/OpenBSD(314)`), no `.Xresources` or `.Xdefaults` used
- tmux from an OpenBSD snapshot dated July 26th (i.e., the latest version from CVS), no `tmux.conf` used

-------

The informal test involved opening a terminal taking up a quarter of my screen (I'm using i3-wm, not sure if that matters), then opening tmux if needed, and finally `vim -U NONE ~/.kshrc` or the same for `nvim`. Once open, I just alternated between hitting Page-up and Page-down:
```
st:
- vim WORKS
- nvim WORKS

st + tmux:
- vim WORKS
- nvim BROKEN

xterm:
- vim WORKS
- nvim WORKS

xterm + tmux:
- vim WORKS
- nvim BROKEN
```

`BROKEN` indicates that flickering was found, and it always occured in the right-most terminal column, nowhere else. Besides Page-up/down, moving the cursor to the bottom of the window and then scrolling further down with `j` caused flickering as well, possibly because the majority of the screen was being redrawn.

An additional problem I found just now was that resizing produces a lot of flickering, but only in tmux. This happens in `tmux` using `nvim -u NONE`, and is trigger by resizes of seemingly any magnitude.

----------------

I don't recall getting this issue when running Arch Linux, so my guess is it's something to do with my system's terminfo, either for st, tmux, xterm, or a combination of the three.

**Comment by khalidchawtany** [comment]
Neovim's Completion menu flickers a lot when running inside **tmux**. However, it is smooth outside of it. Enabling *lazyredraw* makes it worse ( Unbearable :angry:  ).

If the completion menu is drawn over an empty area, it does not flicker. However, if it is drawn over other text, then it flickers a lot.

FYI: I have a **patched tmux** to support true color.
```
$ nv --version
NVIM 0.0.0-alpha+201508131220 (compiled Aug 13 2015 18:46:19)
Commit: a94a68145b3c607d1e58f708ded8fe625c9973d5

$ tmux -V
tmux 2.0

$ echo $TERM
xterm-256color   => Same for both
```

**Comment by breerly** [comment]
I'm seeing this as well.

**Comment by Pyrohh** [comment]
> +1 - I'm getting this as well.

need more info than this...

**Comment by breerly** [comment]
During screen start there is no welcome screen - instead, the cursor dances around the screen in a quick, flickering manor. Once this is done, it goes to direclty to normal mode on line 1:

```

1
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
 NORMAL  [No Name]                                                   [unix]  100%    0:  1
```

**Comment by breerly** [comment]
![tmux-neovim-issue](https://cloud.githubusercontent.com/assets/263587/12827938/edb4d10c-cb36-11e5-8568-ae5a3a387df2.gif)

**Comment by breerly** [comment]
The larger the window size is - the longer and more apparent the flickering becomes.

**Comment by justinmk** [comment]
> During screen start there is no welcome screen 

Sounds like https://github.com/neovim/neovim/issues/3416. Can you please narrow it down to a plugin?

**Comment by breerly** [comment]
With `nvim -u NONE`, it works even inside tmux.

**Comment by justinmk** [comment]
Can you bisect your vimrc to narrow it down to a specific plugin?

**Comment by breerly** [comment]
Hmmm... removing these lines stops the flickering:

```
set statusline +=%#warningmsg#
set statusline+=%{SyntasticStatuslineFlag()}
set statusline+=%*
```

But it's still skipping the welcome screen.

I'll dig deeper.

**Comment by breerly** [comment]
Weird... once I get rid of the following, I get the welcome screen again:

```
set relativenumber       " Relative line numbering
set number               " Display absolute line number instead of 0
```

**Comment by breerly** [comment]
I'm removing config one by one to figure out what causes the flickering...

**Comment by Pyrohh** [comment]
Thanks for looking into it more @breerly. I managed to narrow down the problem further, at least on my system.

---

Version information

- `NVIM v0.1.2-462-gadd02b6`
- tmux from an OpenBSD snapshot dated January 29th, 2016 (i.e., the latest version from CVS)
- st, commit [`6e70bb9`](http://git.suckless.org/st/commit/?id=6e70bb97d4e98a5e2fa7c5488b0cba1da140b78f), with argument `-f terminus-13`

`tmux.conf`:

    set -g default-terminal "screen-256color"

`init.vim`:

```vim
call plug#begin()
Plug 'romainl/flattened'
call plug#end()
colorscheme flattened_dark
```

With the instructions above, I encountered flickering while scrolling downwards through `src/nvim/main.c` using either `}` or `CTRL-D`. When scrolling upwards (with `{` or `CTRL-U`), I don't notice any flickering in the right hand column (see OPs video). Scroll speed doesn't seem to matter; for example, hitting `CTRL-D` once should cause a flicker.

I'll look into it more later, but it seems the issue might be related to highlighting... or something.

**Comment by breerly** [comment]
It looks like removing `vim-airline/vim-airline` and `altercation/vim-colors-solarized` fixes the issue.

**Comment by Pyrohh** [comment]
also see https://marc.info/?l=ncurses-bug&m=129479055132108&w=2

**Comment by Pyrohh** [comment]
I was able to reduce my `init.vim` testcase to one line (taken from the top of `flattened_dark`):
```vim
hi Normal ctermfg=12 ctermbg=8 guifg=#839496 guibg=#002b36 gui=NONE
```

Not really sure what that entails; I'll read the docs on highlighting, but I'd appreciate if some vimscript expert could provide insight.

**Comment by justinmk** [comment]
I watched OP video. @Pyrohh @breerly is this strictly regarding the cursor flickering, or also background color flickering? This just seems like redraw quirks to me. Certainly the cursor flickering has been present for me since the first TUI refactor...

**Comment by Pyrohh** [comment]
> is this strictly regarding the cursor flickering

Yes

> This just seems like redraw quirks to me. Certainly the cursor flickering has been present for me since the first TUI refactor... 

I noticed it at that time too; regardless, it's annoying :grin:

**Comment by justinmk** [comment]
This is probably the same issue as https://github.com/neovim/neovim/issues/4210 which nicely outlines the conditions. When running in tmux, try `ctermbg=NONE` in your colorscheme (for the `Normal` highlight group) and `set noshowcmd`.

**Comment by breerly** [comment]
@justinmk those settings had no effect, still experiencing the issue.

**Comment by mudox** [comment]
Just want to confirm that `hi Normal ctermbg=NONE` did save me out of the long standing startup flickering problem, though it seems like a workaround.
I put it in an appropriate place in my startup color file, and the flickering that lasts for almost one year has gone.

**Comment by kreedz** [comment]
>hi Normal ctermbg=NONE guibg=NONE
But this breaks colorscheme, it gives background from console.
I think glitching remaining anyway, you just don't see it because glitching color now the same as console background.

So I have flickering in neovim (inside tmux) when scrolling.
Outside of tmux it's normal.
tmux 2.3
neovim 0.1.7
urxvt 9.20
alacritty 0.1.0

**Comment by TangMonk** [comment]
same problem

**Comment by tuannvm** [comment]
The `set termguicolors` caused problem for me. Disable it and the flickering's gone

**Comment by mcgaffin** [comment]
@tuannvm Thanks for posting this comment! I have been experiencing this problem for a couple months now and struggled to find a solution. Commenting out `set termguicolors` in my `init.vm` fixed the problem right away.

**Comment by casperstorm** [comment]
I have exactly the same issue with flickering inside tmux, it does fix it with `set termguicolors` but what is going on here?

Running tmux `3.0a` and nvim `0.4.3`. Tested, and have same error with the following terminals:
```
kitty 0.16.0
alacritty 0.4.1
iterm2  3.3.9
```

**Comment by xarthna** [comment]
I have isolated this issue to guicursor blinking settings.  This was the offending setting `a:blinkwait700-blinkoff400-blinkon250-Cursor/lCursor` part of `set guicursor`. After removing that the issue has resolved for me.

OS: Arch 5.14.8-arch1-1
Tmux: 3.2a
NVIM: v0.5.1
XTerm: 369
